File: examples/dekita/.claude/scripts/ci_monitor/events.py
"""Event emission and logging for ci-monitor.

This module handles MonitorEvent creation, emission, and general logging.
Extracted from ci-monitor.py as part of Issue #1754 refactoring.
"""

from __future__ import annotations

import json
import sys
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

from ci_monitor.models import EventType, MonitorEvent

if TYPE_CHECKING:
    from collections.abc import Callable


def emit_event(event: MonitorEvent) -> None:
    """Emit an event to stdout in JSON format.

    Args:
        event: The MonitorEvent to emit.
    """
    print(event.to_json(), flush=True)


def create_event(
    event_type: EventType,
    pr_number: str,
    message: str,
    details: dict[str, Any] | None = None,
    suggested_action: str = "",
    log_background_fn: Callable[[str, str, dict[str, Any]], None] | None = None,
) -> MonitorEvent:
    """Create a MonitorEvent with current timestamp.

    Issue #1663: Also logs the event to background task logger for persistence.

    Args:
        event_type: The type of event (from EventType enum).
        pr_number: The PR number this event relates to.
        message: Human-readable message describing the event.
        details: Optional dictionary with additional event details.
        suggested_action: Optional suggested action for the user.
        log_background_fn: Optional callback to log to background task logger.
            Signature: (task_name, event_type, details) -> None

    Returns:
        A new MonitorEvent instance with current timestamp.
    """
    event = MonitorEvent(
        event_type=event_type,
        pr_number=pr_number,
        timestamp=datetime.now(UTC).isoformat(),
        message=message,
        details=details or {},
        suggested_action=suggested_action,
    )

    # Issue #1663: Log to background task logger for persistence
    if log_background_fn:
        try:
            log_background_fn(
                "ci-monitor",
                event_type.value,
                {
                    "pr_number": pr_number,
                    "message": message,
                    **(details or {}),
                },
            )
        except Exception as e:  # noqa: BLE001 - Don't interrupt monitoring
            # Log warning for debugging - don't interrupt monitoring
            print(f"Warning: Failed to log background event: {e}", file=sys.stderr)

    return event


def log(message: str, json_mode: bool = False, data: dict[str, Any] | None = None) -> None:
    """Print a log message.

    Args:
        message: The message to log.
        json_mode: If True, output as JSON to stderr. If False, output to stdout.
        data: Optional additional data to include in JSON output.
    """
    # Intentionally using local time for user-facing console output.
    # Users expect to see timestamps in their local timezone (e.g., "[03:12:12]").
    timestamp = datetime.now().strftime("%H:%M:%S")
    if json_mode:
        log_data: dict[str, Any] = {"timestamp": timestamp, "message": message, "type": "log"}
        if data:
            log_data.update(data)
        print(json.dumps(log_data, ensure_ascii=False), file=sys.stderr, flush=True)
    else:
        print(f"[{timestamp}] {message}", flush=True)

File: examples/dekita/.claude/scripts/ci_monitor_ts/events.ts
/**
 * Event emission and logging for ci-monitor.
 *
 * Why:
 *   Provide consistent event handling and logging for CI monitoring:
 *   - Structured JSON event emission
 *   - Human-readable console logging
 *   - Background task logger integration
 *
 * What:
 *   - emitEvent(): Emit event to stdout in JSON format
 *   - createEvent(): Create MonitorEvent with current timestamp
 *   - log(): Print log message (JSON or text format)
 *
 * Remarks:
 *   - Migrated from Python ci_monitor/events.py (Issue #3261)
 *
 * Changelog:
 *   - silenvx/dekita#3261: TypeScript migration from Python
 */

import type { EventType, MonitorEvent } from "../../hooks/lib/types";

// =============================================================================
// Types
// =============================================================================

/** Background task logger callback signature */
export type BackgroundLogFn = (
  taskName: string,
  eventType: string,
  details: Record<string, unknown>,
) => void;

// =============================================================================
// Event Functions
// =============================================================================

/**
 * Convert MonitorEvent to dictionary format.
 *
 * @param event - The event to convert
 * @returns Dictionary representation of the event
 */
export function eventToDict(event: MonitorEvent): Record<string, unknown> {
  return {
    event: event.eventType,
    pr_number: event.prNumber,
    timestamp: event.timestamp,
    message: event.message,
    details: event.details,
    suggested_action: event.suggestedAction,
  };
}

/**
 * Convert MonitorEvent to JSON string.
 *
 * @param event - The event to convert
 * @returns JSON string representation of the event
 */
export function eventToJson(event: MonitorEvent): string {
  return JSON.stringify(eventToDict(event), null, 2);
}

/**
 * Emit an event to stdout in JSON format.
 *
 * @param event - The MonitorEvent to emit
 */
export function emitEvent(event: MonitorEvent): void {
  console.log(eventToJson(event));
}

/**
 * Create a MonitorEvent with current timestamp.
 *
 * Issue #1663: Also logs the event to background task logger for persistence.
 *
 * @param eventType - The type of event (from EventType enum)
 * @param prNumber - The PR number this event relates to
 * @param message - Human-readable message describing the event
 * @param details - Optional dictionary with additional event details
 * @param suggestedAction - Optional suggested action for the user
 * @param logBackgroundFn - Optional callback to log to background task logger
 * @returns A new MonitorEvent instance with current timestamp
 */
export function createEvent(
  eventType: EventType,
  prNumber: string,
  message: string,
  details?: Record<string, unknown>,
  suggestedAction?: string,
  logBackgroundFn?: BackgroundLogFn,
): MonitorEvent {
  const event: MonitorEvent = {
    eventType,
    prNumber,
    timestamp: new Date().toISOString(),
    message,
    details: details ?? {},
    suggestedAction: suggestedAction ?? "",
  };

  // Issue #1663: Log to background task logger for persistence
  if (logBackgroundFn) {
    try {
      logBackgroundFn("ci-monitor", eventType, {
        pr_number: prNumber,
        message,
        ...(details ?? {}),
      });
    } catch (error) {
      // Don't interrupt monitoring - just warn
      console.error(`Warning: Failed to log background event: ${error}`);
    }
  }

  return event;
}

/**
 * Print a log message.
 *
 * @param message - The message to log
 * @param jsonMode - If true, output as JSON to stderr. If false, output to stdout
 * @param data - Optional additional data to include in JSON output
 */
export function log(message: string, jsonMode = false, data?: Record<string, unknown>): void {
  // Intentionally using local time for user-facing console output.
  // Users expect to see timestamps in their local timezone (e.g., "[03:12:12]").
  const now = new Date();
  const timestamp = now.toLocaleTimeString("ja-JP", {
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  });

  if (jsonMode) {
    const logData: Record<string, unknown> = {
      timestamp,
      message,
      type: "log",
      ...(data ?? {}),
    };
    console.error(JSON.stringify(logData));
  } else {
    console.log(`[${timestamp}] ${message}`);
  }
}
